estimagic/differentiation/generate_steps.py

```diff
@@ -0,0 +1,255 @@
+        if min_steps is not None:
+            base_steps[base_steps < min_steps] = min_steps
+    return base_steps
+
+
+def _set_unused_side_to_nan(x, pos, neg, method, lower_step_bounds, upper_step_bounds):
+    """Set unused side (i.e. pos or neg) to np.nan.
+
+    A side is not used if:
+    - It was not requested due to one sided derivatives.
+    - It was requested but a side switch was better due to bounds.
+
+    This function does not yet guarantee that all bounds are fulfilled. It only switches
+    to the side that has more space to the bound if there is a bound violation.
+
+    Args:
+        x (np.ndarray): 1d array with parameters.
+        pos (np.ndarray): Array with positive steps of shape (n_steps, len(x))
+        neg (np.ndarray): Array with negative steps of shape (n_steps, len(x))
+        method (str): One of ["forward", "backward"]
+        lower_step_bounds (np.ndarray): Lower bounds for steps.
+        upper_step_bounds (np.ndarray): Upper bounds for steps.
+
+    Returns:
+        pos (np.ndarray): Copy of pos with additional NaNs
+        neg (np.ndarray): Copy of neg with additional NaNs
+
+    """
+    pos = pos.copy()
+    neg = neg.copy()
+    better_side = np.where(upper_step_bounds >= -lower_step_bounds, np.ones_like(x), -1)
+    max_abs_step = pos[:, -1]
+    if method == "forward":
+        used_side = np.where(
+            upper_step_bounds >= max_abs_step, np.ones_like(x), better_side
+        )
+    elif method == "backward":
+        used_side = np.where(
+            -lower_step_bounds >= max_abs_step, -np.ones_like(x), better_side
+        )
+    else:
+        raise ValueError("This function only works for forward or backward method.")
+
+    pos[used_side == -1] = np.nan
+    neg[used_side == 1] = np.nan
+    return pos, neg
+
+
+def _rescale_to_accomodate_bounds(
+    base_steps, pos, neg, lower_step_bounds, upper_step_bounds, min_steps
+):
+    """Rescale steps to make them compatible with bounds unless this violates min_steps.
+
+    Args:
+        base_steps (np.ndarray, optional): 1d array of the same length as x.
+            base_steps * scaling_factor is the absolute value of the first (and possibly
+            only) step used in the finite differences approximation of the derivative.
+        pos (np.ndarray): Array with positive steps of shape (n_steps, len(x))
+        neg (np.ndarray): Array with negative steps of shape (n_steps, len(x))
+        lower_step_bounds (np.ndarray): Lower bounds for steps.
+        upper_step_bounds (np.ndarray): Upper bounds for steps.
+        min_steps (np.ndarray): Minimal possible step sizes that can be chosen
+            to accomodate bounds. Needs to have same length as x.
+
+    Returns:
+        pos (np.ndarray): Copy of pos with rescaled steps.
+        neg (np.ndarray): Copy of neg with rescaled steps.
+
+    """
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=RuntimeWarning)
```

Comment:
Whats the reason of the warning?

Reply:
If you call `nanmax` or `nanmin` on an array that only contains NaNs you get a RuntimeWarning.

![image](https://user-images.githubusercontent.com/9536380/76141883-860bb180-6068-11ea-850b-e445fb5a48c3.png)

This would happen a lot with one-sided derivatives.